test-framework/common/src/main/java/io/quarkus/test/common/PathTestHelper.java

```diff
@@ -137,13 +138,22 @@ public static Path getTestClassesLocation(Class<?> testClass) {
         String classFileName = testClass.getName().replace('.', File.separatorChar) + ".class";
         URL resource = testClass.getClassLoader().getResource(fromClassNameToResourceName(testClass.getName()));
 
+        if (resource == null) {
+            throw new IllegalStateException(
+                    "Could not find resource: " + testClass.getName() + " using class loader " + testClass.getClassLoader());
+        }
         if (resource.getProtocol().equals("jar")) {
             try {
                 resource = URI.create(resource.getFile().substring(0, resource.getFile().indexOf('!'))).toURL();
                 return toPath(resource);
             } catch (MalformedURLException e) {
                 throw new RuntimeException("Failed to resolve the location of the JAR containing " + testClass, e);
             }
+        } else if (resource.getProtocol().equals("quarkus")) {
+            // This is loaded with a quarkus classloader, so we can ask it directly
+            QuarkusClassLoader qcl = (QuarkusClassLoader) testClass.getClassLoader();
+            return qcl.getCuratedApplication().getQuarkusBootstrap().getTestClassesLocation();
```

Comment:
This is where we (theoretically) could try `qcl.getCuratedApplication().getApplicationModel().getAppArtifact().getWorkspaceModule().getTestSources().getOutputDir()`

Reply:
As discussed in other comments, this ended up being a 15-liner, so I isolated it to its own method :)